presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -271,13 +272,19 @@ else if (queryStateMachine.getQueryState() == QueryState.STARTING) {
 
     private static void updateQueryOutputLocations(QueryStateMachine queryStateMachine, OutputBufferId rootBufferId, Set<RemoteTask> tasks, boolean noMoreExchangeLocations)
     {
-        Set<URI> bufferLocations = tasks.stream()
-                .map(task -> task.getTaskStatus().getSelf())
-                .map(location -> uriBuilderFrom(location).appendPath("results").appendPath(rootBufferId.toString()).build())
-                .collect(toImmutableSet());
+        Map<URI, TaskId> bufferLocations = tasks.stream()
+                .collect(toImmutableMap(
+                        task -> getBufferLocationForRemoteTask(task, rootBufferId),
+                        task -> task.getTaskId()));
         queryStateMachine.updateOutputLocations(bufferLocations, noMoreExchangeLocations);
     }
 
+    private static URI getBufferLocationForRemoteTask(RemoteTask remoteTask, OutputBufferId rootBufferId)
```

Comment:
Maybe `getBufferOutputLocation`? Don't feel strong though.

Reply:
hmm.. I would just call `BufferLocation` as it's often used :). I originally want to emphasize we are collecting the buffer location from "receiver"'s point of view... But maybe it's redundant :)